sqlalchemy_utils/functions/database.py

```diff
@@ -459,41 +461,54 @@ def sqlite_file_exists(database):
 
     url = copy(make_url(url))
     database, url.database = url.database, None
-    engine = sa.create_engine(url)
+    engine = None
+    dialect_name = url.get_dialect().name
 
-    if engine.dialect.name == 'postgresql':
+    if dialect_name == 'postgresql':
+        ret = False
         text = "SELECT 1 FROM pg_database WHERE datname='%s'" % database
-        return bool(get_scalar_result(engine, text))
+        if postgres_db is None:
+            postgres_db = ('postgres', 'template0', 'template1', None)
+        for pdb in postgres_db:
+            url.database = pdb
+            engine = sa.create_engine(url)
+            try:
+                ret = bool(get_scalar_result(engine, text))
+                break
+            except (ProgrammingError, OperationalError):
+                pass
+            finally:
+                engine.dispose()
```

Comment:
Do we need to set engine to None here -- is it safe to call dispose() twice on an engine? A second dispose() will be called on line 510

Reply:
true. how about this https://github.com/kvesteri/sqlalchemy-utils/pull/463/commits/e4815822ba7497afbf38891e8e1ca83b85e9f928

Reply:
I think now we aren't disposing engines between each iteration of the loop if there are not exceptions. I'm not the author of this code, so don't want to offer too many opinions, but this code looks a little brittle in terms of the manual calls to Engine.dispose(). I would just use a context manager here...one that called dispose() in __exit__ and then use `with` everywhere I created an Engine

Reply:
If I understand https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=dispose#connectionless-execution-implicit-execution correctly, the current code uses implicit connections (i.e. calling execute on the engine). This may be changed to 

```
with engine.connect() as connection:
    connection.execute(...)
```

then we might no need to use `dispose` at all https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=dispose#engine-disposal .. 

but I'm not an sqlalchemy expert at all .. but I'm fine with implementing the change if you agree

Reply:
also dispose is called in `get_scalar_result` .. i.e. in the case of a successful existence check dispose is currently called twice... but according to the docs it just closes connections.

Reply:
I would do this:

```

Reply:
Thanks for the detailed explanations. I tried a slightly different way in https://github.com/kvesteri/sqlalchemy-utils/pull/463/commits/6b5cdb252c55d138e1335d6b64fcc3a1381681a8 

1. use a connection for the execution and close the connection (via `with`). Already with this there should be no need to dispose the engine (since there should be no open connections in the pool that need to be closed).
2. explicitly use the Null connection pool .. just to be sure.

Lets see if this passes tests.